### ISSUE TITLE
US INTL keyboard gets messed up between server and client (workaround/fix at bottom)

### ISSUE THREAD

**Issue by thiago-negri** [deskflow/deskflow#7891]
### Project

Deskflow

### Deskflow version number

1.17.1.0

### Build type

Deskflow package (downloaded from Deskflow)

### Operating systems (OS)

- [X] Windows
- [X] macOS
- [ ] Linux (X11)
- [ ] Linux (Wayland)
- [ ] BSD-derived
- [ ] Other (please specify)

### Wayland on Linux

- [ ] I have reviewed the Wayland [known issues](https://github.com/deskflow/deskflow/discussions/7499) and my issue is new
- [ ] I am not using Wayland on Linux

### Signing on macOS

- [ ] I have authorized the app to run on my Mac
- [ ] I am not using macOS

### Continuous build

- [ ] I have tried the latest continuous build and the issue persists
- [ ] I am unable to try the latest continuous build

### OS versions/distros

- Windows 11
- macOS Sonoma 14.7.1

### Deskflow configuration

- Windows 11 server, macOS Sonoma 14.7.1 client

### What steps will reproduce the problem?

1. Set Windows 11 keyboard layout to "Portuguese (Brazil) US INTL"
2. Set macOS keyboard layout to "Brazilian"
3. Start Deskflow on both machines
4. Drag mouse to client
5. Start typing
6. When you press a key that requires a "combo" in INTL layout like `'c` to produce `ç`, the first `'` key press don't do anything at client side
7. If you do `'c` on server, you get a `ç` on client as it was a single key stroke
8. But when you do like `'<space>` to get a quote, you get a backtick (\`) as if it was waiting to use the INTL "combo" at client side
9. So if I do `'<space>a` on client, I get `à`, which is incorrect, I expect `'a`
10. `<shift>'<space>` should produce a double quote (`"`), but again `<shift>'` doesn't seem to register anything at client side, but when I do `<shift>'<space>` I get a tilde `~` at client side waiting for an INTL "combo", so `<shift>'<space>a` becomes `ã` at client side when it should be `"a`

Also tried switching to an ABNT keyboard, setting both server and clients to "Brazil ABNT2" keyboard layouts, but it also messes up keys. As INTL is the main keyboard I want to use (and probably easier to setup in other countries), I'm focusing this issue on that.

### Log output

```shell
-- Server Log when pressing '
[2024-11-13T09:37:58] DEBUG1: hook: 0x000000de 0x00280001
[2024-11-13T09:37:58] DEBUG1: hook: 0x6fde0027 0x00280001
[2024-11-13T09:37:58] DEBUG1: hook: 0x000000de 0x80280001
[2024-11-13T09:37:58] DEBUG1: hook: 0x040000de 0x80280001

-- Server Log when pressing '<SPACE> (required for an actual quote appear, as this is INTL layout)
[2024-11-13T09:38:57] DEBUG1: hook: 0x000000de 0x00280001
[2024-11-13T09:38:57] DEBUG1: hook: 0x6fde0027 0x00280001
[2024-11-13T09:38:57] DEBUG1: hook: 0x000000de 0x80280001
[2024-11-13T09:38:57] DEBUG1: hook: 0x040000de 0x80280001
[2024-11-13T09:38:58] DEBUG1: hook: 0x00000020 0x00390001
[2024-11-13T09:38:58] DEBUG1: hook: 0x61200027 0x00390001
[2024-11-13T09:38:58] DEBUG1: hook: 0x07200027 0x00390001
[2024-11-13T09:38:58] DEBUG1: event: Key char=39, vk=0x20, nagr=0, lParam=0x00390001
[2024-11-13T09:38:58] DEBUG1: new mask: 0x2000
[2024-11-13T09:38:58] DEBUG1: new mask: 0x2000
[2024-11-13T09:38:58] DEBUG1: onKeyDown id=39 mask=0x2000 button=0x0039 lang=pt
[2024-11-13T09:38:58] DEBUG1: hook: 0x00000020 0x80390001
[2024-11-13T09:38:58] DEBUG1: hook: 0x61200020 0x80390001
[2024-11-13T09:38:58] DEBUG1: hook: 0x07200020 0x80390001
[2024-11-13T09:38:58] DEBUG1: event: Key char=32, vk=0x20, nagr=0, lParam=0x80390001
[2024-11-13T09:38:58] DEBUG1: new mask: 0x2000
[2024-11-13T09:38:58] DEBUG1: new mask: 0x2000
[2024-11-13T09:38:58] DEBUG1: onKeyUp id=32 mask=0x2000 button=0x0039

-- Client Log when pressing '
.... nothing ....

-- Client Log when pressing '<SPACE> (required for an actual quote appear, as this is INTL layout)
[2024-11-13T09:40:55] DEBUG1: recv key down id=0x00000027, mask=0x2000, button=0x0039, lang="pt"
[2024-11-13T09:40:55] DEBUG1: mapKey 0027 (39) with mask 2000, start state: 0000, group: 1
[2024-11-13T09:40:55] DEBUG1: language pt has group id 1
[2024-11-13T09:40:55] DEBUG1: best key index 1 of 1 (exact)
[2024-11-13T09:40:55] DEBUG1: found key in group 2
[2024-11-13T09:40:55] DEBUG1: language pt has group id 1
[2024-11-13T09:40:55] DEBUG1: state: 0000,0000,0007
[2024-11-13T09:40:55] DEBUG1: flip: 0000 (0000 vs 0000 in 0007 - 0000)
[2024-11-13T09:40:55] DEBUG1: desired state: 2000 0000,0000,0007
[2024-11-13T09:40:55] DEBUG1: flip: 0000 (0000 vs 2000 in fff8 - 6020)
[2024-11-13T09:40:55] DEBUG1: mapped to 033, new state 0000
[2024-11-13T09:40:55] DEBUG1: keystrokes:
[2024-11-13T09:40:55] DEBUG1:   group 1
[2024-11-13T09:40:55] DEBUG1: keyboard layout change to 1
[2024-11-13T09:40:55] DEBUG1:   button=0x0033 virtualKey=0x0032 keyDown=down client=0x0000
[2024-11-13T09:40:55] DEBUG1: the key is not a modifier
[2024-11-13T09:40:55] DEBUG1: recv key up id=0x00000027, mask=0x2000, button=0x0039
[2024-11-13T09:40:55] DEBUG1: keystrokes:
[2024-11-13T09:40:55] DEBUG1:   button=0x0033 virtualKey=0x0032 keyDown=up client=0x0000
[2024-11-13T09:40:55] DEBUG1: the key is not a modifier
```

### Additional information

I was trying Synergy first, and noticed the issue. Switched to Deskflow to see if it would help, but I got the same behavior.

From the logs it seems that the first key down on a "special" key of an INTL keyboard layout is not sent over the wire by Deskflow, but it detects something at server side (as it logs a bunch of `hook` messages).

Tried restarting Deskflow on both server and client, tried restarting the computers, tried setting up different layouts like US INTL at server, US on client, vice-versa, nothing seems to get both the server and the client to work in the same way.

Even if I set both keyboard layouts to US, when I press the quote (') on the server, the client gets a backtick (\`).

-------

As I was writing this bug report, I noticed that when I set the macOS layout to US and I try to type on it, it automatically switches to Brazilian (INTL).

I went ahead and removed all keyboard layouts from my macOS, leaving it only to default US.

Set my server to US INTL, while macOS is now on US only. Now Deskflow is emitting a warning that it can't set the layout on client. Then I restarted both Deskflow processes.

And suddenly it's working correctly. So I can use Deskflow! The client do not get the first keydown from server, but it gets the quotes and INTL letters correctly after the "combo" is performed.

As I've already written this entire bug report, and this still seems to be an issue that could be prevented by Deskflow itself or documented on how to properly configure it, as I found a workaround by luck. If you don't think so, at least this leaves a workaround for someone else going through the same problem.

Other issues of the same nature:
- #4280
- #5496

**Comment by nbolton** [comment]
Nice workaround! I don't suppose you'd mind trying this workaround in Synergy too? If it's not too much trouble. My expectation is that it will behave the same.

**Comment by thiago-negri** [comment]
@nbolton I'm back to using Synergy, and it works fine.

**Comment by sithlord48** [comment]
yeah we are not doing something proper for the layout / language department.

I suspect this code needs to be looked in at with i18n in mind. Chances are this code was written for ascii only and all the "support" was hacked on later and clearly has suffered some bit rot since.

**Comment by SGudbrandsson** [comment]
I have my computers with two keyboard layouts which I swap out frequently, Icelandic and US.
The keyboard characters sent over are too messed up to develop code reliably, forcing me to have two keyboards.

One example is when I have the Icelandic keyboard layout on both computers and want to type the `@` symbol.
Normally, I would type `alt`+`q`, however, that just gives me ²
If I try US on both computers, it all turns to weird keypresses.

What I need to do is to set the keyboard layout to IS on the server, and US on the client, then I can use `shift`+`2` to send an `@` symbol.
I have a similar situation when trying to type `:` with `shift`+`æ` (character right of `L`). It turns the keypress into X for some reason.

@sithlord48 I am willing to help debug this if needed.
Do you need help debugging?

**Comment by sithlord48** [comment]
Yes any help here is welcome. I am not affected by this as i have a boring normal us keyboard.

**Comment by MSK61** [comment]
I have the same problem with a server(windows 10) with two keyboard layouts "English US INTL" and "Arabic Egypt" and a client(windows 11) with the same 2 keyboard layouts and one extra layout "German Germany". I set the server to use "English US INTL" and then switch to the client which I also want to use "English US INTL". When I try to type a single quote with the key combo '(quote) followed by SPACE, I get instead a ?(question mark) on the client. If I try to type a double quote with the key combo SHIFT+'(quote) followed by SPACE, I get instead a @(at sign) on the client. I couldn't apply any workarounds in my case as I need the other keyboard layouts.

**Comment by tekknovator** [comment]
For me the workaround is the other way round. I set my ubuntu server to U.S. and my macOS client to U.S.International PC.
This means I loose the gestures on ubuntu.
I will try to get used working the the gnome compose key instead for umlauts and french stuff.
Edit: Also make sure to set the locale correctly.
`localectl set-x11-keymap us; setxkbmap -layout us`
